sinks/sink-mongo/src/sink.rs

```diff
@@ -136,34 +158,92 @@ impl Sink for MongoSink {
+        for collection in self.collections.values() {
+            collection
+                .delete_many_with_session(delete_query.clone(), None, &mut session)
+                .await
+                .change_context(SinkMongoError)
+                .attach_printable("failed to invalidate data (delete)")?;
+            collection
+                .update_many_with_session(
+                    unclamp_query.clone(),
+                    unset_cursor_to.clone(),
+                    None,
+                    &mut session,
+                )
+                .await
+                .change_context(SinkMongoError)
+                .attach_printable("failed to invalidate data (update)")?;
+        }
 
         Ok(())
     }
 }
 
 impl MongoSink {
+    pub fn get_collection(
+        &self,
+        collection_name: &str,
+    ) -> Result<&Collection<Document>, SinkMongoError> {
+        self.collections
+            .get(collection_name)
+            .ok_or(SinkMongoError)
+            .attach_printable(format!("collection '{collection_name}' not found"))
+    }
+
     pub async fn insert_data(
         &self,
         end_cursor: &Cursor,
         values: &[Value],
     ) -> Result<(), SinkMongoError> {
-        let docs = values
+        let mut docs = values
             .iter()
             .map(to_document)
             .collect::<std::result::Result<Vec<_>, _>>()
             .change_context(SinkMongoError)
             .attach_printable("failed to convert batch to mongo document")?;
 
+        let mut docs_map: HashMap<String, Vec<Document>> = HashMap::new();
+
+        if self.collections.len() > 1 {
+            for doc in docs.iter_mut() {
+                let collection_name = doc
+                    .remove("collection")
+                    .ok_or(SinkMongoError)
+                    .attach_printable("document missing collection key")?
+                    .as_str()
+                    .ok_or(SinkMongoError)
+                    .attach_printable("collection is not a string")?
+                    .to_string();
+
+                if let Mode::Logs = self.mode {
+                    *doc = doc
+                        .get("data")
+                        .ok_or(SinkMongoError)
+                        .attach_printable("document missing data key")?
+                        .as_document()
+                        .ok_or(SinkMongoError)
+                        .attach_printable("data is not a document")?
+                        .clone();
+                }
+
+                if !docs_map.contains_key(&collection_name) {
+                    docs_map.insert(collection_name.clone(), Vec::new());
+                }
+
+                // Safe unwrap because of the above if statement where we add the
+                // `collection_name` key if it doesn't exist
+                docs_map
+                    .get_mut(&collection_name)
+                    .unwrap()
+                    .push(doc.clone());
```

Comment:
I wonder if it's replaceable with the [entry pattern](https://doc.rust-lang.org/std/collections/hash_map/enum.Entry.html). Something like 

```rs
docs_map
  .entry(&collection_name)
  .or_default()
  .push(doc.clone())
```